presto-main/src/main/java/com/facebook/presto/operator/scalar/ArraysOverlapFunction.java

```diff
@@ -13,202 +13,105 @@
-            @SqlType("array(E)") Block rightArray)
+    public static Boolean arraysOverlap(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block leftArray,
+            @SqlType("array(T)") Block rightArray)
     {
-        int leftPositionCount = leftArray.getPositionCount();
-        int rightPositionCount = rightArray.getPositionCount();
-
-        if (leftPositionCount == 0 || rightPositionCount == 0) {
+        if (leftArray.getPositionCount() == 0 || rightArray.getPositionCount() == 0) {
             return false;
         }
 
-        if (leftPositions == null || leftPositions.length < leftPositionCount) {
-            leftPositions = new int[leftPositionCount * 2];
+        if (Math.max(rightArray.getPositionCount(), leftArray.getPositionCount()) >= NESTED_LOOP_THRESHOLD) {
+            return arraysOverlapSetBased(elementType, leftArray, rightArray);
         }
 
-        if (rightPositions == null || rightPositions.length < rightPositionCount) {
-            rightPositions = new int[rightPositionCount * 2];
-        }
-
-        for (int i = 0; i < leftPositionCount; i++) {
-            leftPositions[i] = i;
-        }
-        for (int i = 0; i < rightPositionCount; i++) {
-            rightPositions[i] = i;
-        }
-        IntArrays.quickSort(leftPositions, 0, leftPositionCount, intBlockCompare(type, leftArray));
-        IntArrays.quickSort(rightPositions, 0, rightPositionCount, intBlockCompare(type, rightArray));
-
-        int leftCurrentPosition = 0;
-        int rightCurrentPosition = 0;
-        while (leftCurrentPosition < leftPositionCount && rightCurrentPosition < rightPositionCount) {
-            if (leftArray.isNull(leftPositions[leftCurrentPosition]) || rightArray.isNull(rightPositions[rightCurrentPosition])) {
-                // Nulls are in the end of the array. Non-null elements do not overlap.
-                return null;
+        boolean hasNull = false;
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            if (leftArray.isNull(i)) {
+                hasNull = true;
+                continue;
             }
-
-            int compareValue;
             try {
-                compareValue = type.compareTo(leftArray, leftPositions[leftCurrentPosition], rightArray, rightPositions[rightCurrentPosition]);
-            }
-            catch (NotSupportedException e) {
-                throw new PrestoException(NOT_SUPPORTED, e.getMessage(), e);
-            }
-
-            if (compareValue > 0) {
-                rightCurrentPosition++;
-            }
-            else if (compareValue < 0) {
-                leftCurrentPosition++;
+                for (int j = 0; j < rightArray.getPositionCount(); j++) {
+                    if (rightArray.isNull(j)) {
+                        hasNull = true;
+                        continue;
+                    }
+                    if (elementType.equalTo(leftArray, i, rightArray, j)) {
```

Comment:
Simplified the code duplicity by using equalTo() function in Type, since TypedSet also uses that internally.

Also so slight improvement in the performance for small arrays benchmark.

# Run complete. Total time: 00:03:50

Benchmark                              Mode  Cnt   Score   Error  Units
BenchmarkArraysOverlap.largeBenchmark  avgt   20  96.736 ± 3.395  ns/op
BenchmarkArraysOverlap.smallBenchmark  avgt   20   9.419 ± 0.490  ns/op